pkg/kubelet/kubelet.go

```diff
@@ -759,6 +765,61 @@ func (kl *Kubelet) makeEnvironmentVariables(ns string, container *api.Container)
 	return result, nil
 }
 
+// runtimeEnvVarValue determines the value that an env var should take when a container
+// is started.  If the value of the env var is the empty string, the source of the env var
+// is resolved, if one is specified.
+//
+// TODO: preliminary factoring; make better
+func (kl *Kubelet) runtimeEnvVarValue(envVar api.EnvVar, pod *api.Pod) (string, error) {
+	runtimeVal := envVar.Value
+	if runtimeVal != "" {
+		return runtimeVal, nil
+	}
+
+	switch {
+	case envVar.ObjectInfo != nil:
+		return kl.objectInfoRuntimeValue(envVar.ObjectInfo, pod)
+	case envVar.HostInfo != nil:
+		return kl.hostInfoRuntimeValue(envVar.HostInfo)
+	}
+
+	return runtimeVal, nil
+}
+
+func (kl *Kubelet) objectInfoRuntimeValue(objInfo *api.ObjectReference, pod *api.Pod) (string, error) {
+	// TODO: use objInfo Kind instead of "Pod"
+	internalFieldPath, _, err := api.Scheme.ConvertFieldLabel(objInfo.APIVersion, "Pod", objInfo.FieldPath, "")
+	if err != nil {
+		return "", err
+	}
+
+	runtimeVal := ""
+
+	// TODO: currently there is no generic utility for retrieving a field of
+	// an object using a fieldPath expression.  For now, support only the
+	// Name and Namespace fields.
+	switch internalFieldPath {
+	case "metadata.name":
```

Comment:
Is it worthwhile to extract this to a pkg/fieldpath or pkg/api/fieldpath so that we can encapsulate later changes to the spec, and maybe offer validation hooks?